### PR TITLE
Rule: Add No Switch

### PR DESCRIPTION
**Pull request by Distrib-ps**
Pr for https://www.smogon.com/forums/threads/no-switch-rule.3653057/

For this rule, pokemon cannot switch normally.
But he can switch with an attack (teleport, baton pass)...

### CODE REVIEW

File: data/rulesets.ts [hunk 1]
```diff
@@ -1357,4 +1357,13 @@ export const Rulesets: {[k: string]: FormatData} = {
 		desc: "Bans move combinations on Pok\u00e9mon that weren't legal in Japanese versions of Gen 1.",
 		// Implemented in mods/gen1jpn/rulesets.ts
 	},
+	noswitch: {
+		effectType: 'Rule',
+		name: 'No Switch',
+		desc: 'It is not possible to switch Pok&eacute;mon. Except with an item, moove or Ability provided for this purpose.',
```
Comment (DaWoblefet) [inline data/rulesets.ts:1363]:
```suggestion
		desc: 'Players cannot switch their Pok&eacute;mon, except with an item, move, or Ability.',
```

**Comment by DaWoblefet** [comment]
How does this interact with Ghost-types and Shed Shell ignoring trapping effects? Should they?

**Comment by Distrib-ps** [comment]
> How does this interact with Ghost-types and Shed Shell ignoring trapping effects? Should they?
Ghost-types not ignoring effects. But Shed Shell ignoring trapping effects. I don’t know if we should ban  the switch Ability/moves/items of switch.

**Comment by xInstruct** [comment]
> How does this interact with Ghost-types and Shed Shell ignoring trapping effects? Should they?

I 100% think this should be Gen 4 Ingrain trapping. because when I think of a No Switch Mod, I think the button to switch should be hidden entirely.

File: data/rulesets.ts [hunk 2]
```diff
@@ -1357,4 +1357,13 @@ export const Rulesets: {[k: string]: FormatData} = {
 		desc: "Bans move combinations on Pok\u00e9mon that weren't legal in Japanese versions of Gen 1.",
 		// Implemented in mods/gen1jpn/rulesets.ts
 	},
+	noswitch: {
+		effectType: 'Rule',
+		name: 'No Switch',
+		desc: 'Players cannot switch their Pok&eacute;mon, except with an item, move, or Ability.',
+		onTrapPokemon(pokemon) {
+			pokemon.trapped = true;
+			this.add('rule', '-activate', pokemon, 'trapped');
```
Comment (Zarel) [inline data/rulesets.ts:1366]: Uh this is going to look super ugly and also be very spammy (appearing for every pokemon every turn)

Comment (Distrib-ps) [reply data/rulesets.ts]: Done, thx.

File: data/rulesets.ts [hunk 3]
```diff
@@ -1357,4 +1357,15 @@ export const Rulesets: {[k: string]: FormatData} = {
 		desc: "Bans move combinations on Pok\u00e9mon that weren't legal in Japanese versions of Gen 1.",
 		// Implemented in mods/gen1jpn/rulesets.ts
 	},
+	noswitch: {
+		effectType: 'Rule',
+		name: 'No Switch',
```
Comment (Zarel) [inline data/rulesets.ts:1362]: Perhaps "No Switching" would be a better name?

Comment (Zarel) [reply data/rulesets.ts:1362]: I'll ask.

File: data/rulesets.ts [hunk 4]
```diff
@@ -1357,4 +1357,15 @@ export const Rulesets: {[k: string]: FormatData} = {
...
+		desc: 'Players cannot switch their Pok&eacute;mon, except with an item, move, or Ability.',
+		onBegin() {
+			this.add('rule', 'No Switch : Players cannot switch their Pok&eacute;mon');
```
Comment (Zarel) [inline data/rulesets.ts:1365]:
```suggestion
			this.add('rule', 'No Switch: All Pok\u00e9mon are trapped');
```

Comment (Zarel) [reply data/rulesets.ts]: Let me guess, you're French? Is that the language that has the space before `:`? Anyway, in English, we don't.

Comment (Distrib-ps) [reply data/rulesets.ts]: > Let me guess, you're French? Is that the language that has the space before `:`? Anyway, in English, we don't.

Yes! Absolutely. I didn’t know.

File: data/rulesets.ts [hunk 5]
```diff
@@ -1357,4 +1357,15 @@ export const Rulesets: {[k: string]: FormatData} = {
 		desc: "Bans move combinations on Pok\u00e9mon that weren't legal in Japanese versions of Gen 1.",
 		// Implemented in mods/gen1jpn/rulesets.ts
 	},
+	noswitch: {
+		effectType: 'Rule',
+		name: 'No Switch',
+		desc: 'Players cannot switch their Pok&eacute;mon, except with an item, move, or Ability.',
```
Comment (Zarel) [inline data/rulesets.ts:1363]:
```suggestion
		desc: 'All Pok\u00e9mon are trapped (cannot switch naturally, but can as the effect of an item, move, or Ability).',
```

**Comment by Zarel** [comment]
image

**Comment by Distrib-ps** [comment]
Done @Zarel

**Comment by Zarel** [comment]
Thanks!

**Pull request merged**